src/drive/mobile/modules/offline/duck.js

```diff
@@ -0,0 +1,112 @@
+  switch (action.type) {
+    case MAKE_AVAILABLE_OFFLINE:
+      return [...state, action.id]
+    case UNDO_MAKE_AVAILABLE_OFFLINE:
+      const index = state.indexOf(action.id)
+      return [...state.slice(0, index), ...state.slice(index + 1)]
+    default:
+      return state
+  }
+}
+
+const markAsAvailableOffline = id => ({ type: MAKE_AVAILABLE_OFFLINE, id })
+const markAsUnavailableOffline = id => ({
+  type: UNDO_MAKE_AVAILABLE_OFFLINE,
+  id
+})
+
+export const getAvailableOfflineIds = ({ availableOffline }) => availableOffline
+
+export const isAvailableOffline = ({ availableOffline: state }, id) =>
+  Array.isArray(state) && state.indexOf(id) !== -1
+
+export const toggleAvailableOffline = file => async (dispatch, getState) =>
+  isAvailableOffline(getState(), file.id)
+    ? dispatch(forgetDownloadedFile(file))
+    : dispatch(makeAvailableOffline(file))
+
+const forgetDownloadedFile = file => async dispatch => {
+  const filename = file.id
+  if (isCordova() && window.cordova.file) {
+    deleteOfflineFile(filename)
+  }
+  dispatch(markAsUnavailableOffline(file.id))
+}
+
+const makeAvailableOffline = file => async dispatch => {
+  await saveOfflineFileCopy(file)
+  dispatch(markAsAvailableOffline(file.id))
+}
+
+const saveOfflineFileCopy = async file => {
+  if (!isCordova() || !window.cordova.file) {
+    return
+  }
+  const response = await cozy.client.files
+    .downloadById(file.id)
+    .catch(error => {
+      Alerter.error('mobile.error.make_available_offline.offline')
+      throw error
+    })
+  const blob = await response.blob()
+  const filename = file.id
+  saveFileWithCordova(blob, filename)
+}
+
+export const openLocalFile = file => async (dispatch, getState) => {
+  if (!isAvailableOffline(getState(), file.id)) {
+    console.error('openLocalFile: this file is not available offline')
+  }
+  openOfflineFile(file).catch(error => {
+    console.error('openLocalFile', error)
+    Alerter.error('mobile.error.make_available_offline.noapp')
+  })
+}
+
+export const getLocalFileCopyUrl = file => async (dispatch, getState) => {
+  if (isAvailableOffline(getState(), file.id)) {
+    const localFile = await getNativeFile(file)
+    return localFile.nativeURL
+  }
+  const response = await cozy.client.files.downloadById(file.id)
+  const blob = await response.blob()
+  const localFile = await temporarySave(blob, file.name)
+  return localFile.nativeURL
+}
+
+export const openLocalFileCopy = file => async (dispatch, getState) => {
+  if (isAvailableOffline(getState(), file.id)) {
+    return openOfflineFile(file)
+  }
+  const localFile = await createTemporaryLocalFile(file.id, file.name)
+  return openFileWithCordova(localFile.nativeURL, file.mime)
+}
+
+export const updateOfflineFileCopyIfNecessary = (file, prevFile) => async (
+  _,
+  getState
+) => {
+  if (
+    isAvailableOffline(getState(), file.id) &&
+    file.md5sum !== prevFile.md5sum
+  ) {
+    await saveOfflineFileCopy(file)
+  }
+}
```

Comment:
It would have been helpful to move the file in one commit and do the edits in a different one — as things are I can't tell what has changed and what has just been moved 🤷🏼‍♂️

`updateOfflineFileCopyIfNecessary ` is really nice and simple. I don't like it too much that we have functions like `openLocalFileCopy ` in a `duck` file even though they have no interaction with redux.

Reply:
> It would have been helpful to move the file in one commit and do the edits in a different one 

Yes, sorry for that, I didn't think I'd end up with so many changes, I should have been more methodic.

I tried to put every "actions" relative to local file copies because the reducer is very simple right now but we could add a lot of more state (like for instance the status of a file download, aka is it really available?) and make the offline feature better by enhancing these actions with redux state.
This particular function could go in `lib/filesystem.js`, but I think there's more refactoring than just that to do in this part of the code. No religion here, just a tentative arrangement of functions.

Reply:
Yup, I figured it'd be something like that. It's still an improvement over what it was before :)